modules/capability/capability_test.go

```diff
@@ -0,0 +1,104 @@
+package capability_test
+
+import (
+	"testing"
+
+	abci "github.com/cometbft/cometbft/abci/types"
+	tmproto "github.com/cometbft/cometbft/proto/tendermint/types"
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/module"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	ibcmock "github.com/cosmos/ibc-go/v7/testing/mock"
+	"github.com/cosmos/ibc-go/v7/testing/simapp"
+	"github.com/stretchr/testify/suite"
+
+	"github.com/cosmos/ibc-go/modules/capability"
+	"github.com/cosmos/ibc-go/modules/capability/keeper"
+	"github.com/cosmos/ibc-go/modules/capability/types"
+)
+
+type CapabilityTestSuite struct {
+	suite.Suite
+	cdc    codec.Codec
+	ctx    sdk.Context
+	app    *simapp.SimApp
+	keeper *keeper.Keeper
+	module module.AppModule
+}
+
+func (suite *CapabilityTestSuite) SetupTest() {
+	checkTx := false
+	app := simapp.Setup(checkTx)
+	cdc := app.AppCodec()
+
+	// create new keeper so we can define custom scoping before init and seal
+	keeper := keeper.NewKeeper(cdc, app.GetKey(types.StoreKey), app.GetMemKey(types.MemStoreKey))
+
+	suite.app = app
+	suite.ctx = app.BaseApp.NewContext(checkTx, tmproto.Header{Height: 1})
+	suite.keeper = keeper
+	suite.cdc = cdc
+	suite.module = capability.NewAppModule(cdc, *keeper, false)
+}
+
+// The following test case mocks a specific bug discovered in https://github.com/cosmos/cosmos-sdk/issues/9800
+// and ensures that the current code successfully fixes the issue.
+func (suite *CapabilityTestSuite) TestInitializeMemStore() {
+	sk1 := suite.keeper.ScopeToModule(banktypes.ModuleName)
+
+	cap1, err := sk1.NewCapability(suite.ctx, "transfer")
+	suite.Require().NoError(err)
+	suite.Require().NotNil(cap1)
+
+	// mock statesync by creating new keeper that shares persistent state but loses in-memory map
+	newKeeper := keeper.NewKeeper(suite.cdc, suite.app.GetKey(types.StoreKey), suite.app.GetMemKey(ibcmock.MemStoreKey))
+	newSk1 := newKeeper.ScopeToModule(banktypes.ModuleName)
+
+	// Mock App startup
+	ctx := suite.app.BaseApp.NewUncachedContext(false, tmproto.Header{})
+	newKeeper.Seal()
+	suite.Require().False(newKeeper.IsInitialized(ctx), "memstore initialized flag set before BeginBlock")
+
+	// Mock app beginblock and ensure that no gas has been consumed and memstore is initialized
+	ctx = suite.app.BaseApp.NewContext(false, tmproto.Header{}).WithBlockGasMeter(sdk.NewGasMeter(50))
+
+	prevBlockGas := ctx.BlockGasMeter().GasConsumed()
+	prevGas := ctx.BlockGasMeter().GasConsumed()
+
```

Comment:
merge conflict?

Reply:
this was done by hand to fix the tests to make them use the old wiring (non-appv2)

Reply:
I'm a bit confused? `prevBlockGas` and `prevGas` look like duplicates to me?

Reply:
```
suite.Require().Equal(prevBlockGas, blockGasUsed, "ensure beginblocker consumed no block gas during execution")
suite.Require().Equal(prevGas, gasUsed, "ensure beginblocker consumed no gas during execution")
```

Reply:
same with `blockGasUsed` and `gasUsed`:
```go
gasUsed := ctx.GasMeter().GasConsumed()

	suite.Require().True(newKeeper.IsInitialized(ctx), "memstore initialized flag not set")
blockGasUsed := ctx.BlockGasMeter().GasConsumed()
```